redbot/cogs/audio/apis/spotify.py

```diff
@@ -0,0 +1,186 @@
+
+    async def get_token(self) -> None:
+        """Get the stored spotify tokens"""
+        if not self._token:
+            self._token = await self.bot.get_shared_api_tokens("spotify")
+
+        self.client_id = self._token.get("client_id", "")
+        self.client_secret = self._token.get("client_secret", "")
+
+    async def get_country_code(self, ctx: Context = None) -> str:
+        return await self.config.guild(ctx.guild).country_code() if ctx else "US"
+
+    async def request_access_token(self) -> MutableMapping:
+        """Make a spotify call to get the auth token"""
+        await self.get_token()
+        payload = {"grant_type": "client_credentials"}
+        headers = self.make_auth_header(self.client_id, self.client_secret)
+        r = await self.post(TOKEN_ENDPOINT, payload=payload, headers=headers)
+        return r
+
+    async def get_access_token(self) -> Optional[str]:
+        """Get the access_token"""
+        if self.spotify_token and not await self.is_access_token_valid(self.spotify_token):
+            return self.spotify_token["access_token"]
+        token = await self.request_access_token()
+        if token is None:
+            log.debug("Requested a token from Spotify, did not end up getting one.")
+        try:
+            token["expires_at"] = int(time.time()) + int(token["expires_in"])
+        except KeyError:
+            return None
+        self.spotify_token = token
+        log.debug(f"Created a new access token for Spotify: {token}")
+        return self.spotify_token["access_token"]
+
+    async def post(
+        self, url: str, payload: MutableMapping, headers: MutableMapping = None
+    ) -> MutableMapping:
+        """Make a POST call to spotify"""
+        async with self.session.post(url, data=payload, headers=headers) as r:
+            data = await r.json()
+            if r.status != 200:
+                log.debug(f"Issue making POST request to {url}: [{r.status}] {data}")
+            return data
+
+    async def make_get_call(self, url: str, params: MutableMapping) -> MutableMapping:
+        """Make a Get call to spotify"""
+        token = await self.get_access_token()
+        return await self.get(url, params=params, headers={"Authorization": f"Bearer {token}"})
+
+    async def get_categories(self, ctx: Context = None) -> List[MutableMapping]:
+        """Get the spotify categories"""
+        country_code = await self.get_country_code(ctx=ctx)
+        params: MutableMapping = {"country": country_code} if country_code else {}
+        result = await self.make_get_call(CATEGORY_ENDPOINT, params=params)
+        with contextlib.suppress(KeyError):
+            if result["error"]["status"] == 401:
+                raise SpotifyFetchError(
+                    message=(
+                        "The Spotify API key or client secret has not been set properly. "
+                        "\nUse `{prefix}audioset spotifyapi` for instructions."
+                    )
```

Comment:
THIS IS PART 1/8 THAT NEED TO BE APPLIED TOGETHER WITH THIS.

```suggestion
                    message=_(
                        "The Spotify API key or client secret has not been set properly. "
                        "\nUse `{prefix}audioset spotifyapi` for instructions."
                    )
```